File: DSA/02Array/Assignments/13_excellent_pairs/excellent_pairs.py
#Link: https://leetcode.com/problems/number-of-excellent-pairs/description/

# class Solution:
#     def countExcellentPairs(self, nums: List[int], k: int) -> int:
#         c = Counter(map(int.bit_count, set(nums)))
#         return sum(c[k1] * c[k2] for k1 in c for k2 in c if k1 + k2 >= k)

class Solution:
    def countExcellentPairs(self, nums: List[int], k: int) -> int:
        c = Counter(map(lambda n:n.bit_count(), set(nums)))
        return sum(c[k1] * c[k2] for k1 in c for k2 in c if k1 + k2 >= k)

File: DSA/02Array/Assignments/13_excellent_pairs/excellent_pairs.cpp
class Solution {
public:
    long long countExcellentPairs(vector<int>& nums, int k) {
    long long cnt[30] = {}, res = 0;
    for (int n : unordered_set<int>(begin(nums), end(nums)))
        ++cnt[__builtin_popcount(n)];
    for (int i = 1; i < 30; ++i)
        for (int j = max(i, k - i); j < 30; ++j)
            res += cnt[i] * cnt[j] * (i == j ? 1 : 2);
    return res;
}
};